internal/database/basestore/handle.go

```diff
@@ -153,3 +166,70 @@ func makeSavepointID() (string, error) {
 
 	return fmt.Sprintf("sp_%s", strings.ReplaceAll(id.String(), "-", "_")), nil
 }
+
+var ErrConcurrentTransactionAccess = errors.New("transaction used concurrently")
+
+// lockingTx wraps a *sql.Tx with a mutex, and reports when a caller tries to
+// use the transaction concurrently. Since using a transaction concurrently is
+// unsafe, we want to catch these issues. If lockingTx detects that a
+// transaction is being used concurrently, it will log an error and attempt to
+// serialize the transaction accesses.
+// NOTE: this is not foolproof. Interleaving savepoints, accessing rows while
```

Comment:
```suggestion
// serialize the transaction accesses.
//
// NOTE: this is not foolproof. Interleaving savepoints, accessing rows while
```

Just to make it look good, otherwise

<img width="150" alt="CleanShot 2022-06-15 at 11 05 34@2x" src="https://user-images.githubusercontent.com/2946214/173728118-96ba4c3d-7f80-4bf4-a87f-dc68e6199eaf.png">